### PR TITLE
refactor(observer): replace chainName by chainID for ChainNonces indexing

### PR DESCRIPTION
**Pull request by lumtis**
# Description

Closes: https://github.com/zeta-chain/node/issues/1490

Replace the indexing for the ChainNonces object.

While this is a logical change since chain ID is the unique identifier for a chain and we should remove the chain name field over time, this change is necessary for https://github.com/zeta-chain/node/issues/2005 because ChainName is a field for static chain information and it is a enum representing hardcoded values. This PR remove the dependency to this ChainName attribute, makes the whole chain information indexed by the chain ID that can be dynamically added.

Over time we will completely remove chain name enum field.

- Change the indexing from chain name to chain ID
- Add a migration script for observer module updating all the indexes for chainNonces object

# How Has This Been Tested?

- Unit test updated
- Migration script with unit test implemented
- Upgrade test passed, meaning the nonces are used with the previous indexing in test, then used with the new indexing

## Summary by CodeRabbit

- **New Features**
	- Introduced an observer module that enhances the application’s capabilities for monitoring events.
	- The API endpoints for querying chain nonces have been updated to use `chain_id` instead of `index`.
	- Added migration functionality to transition data from version 7 to version 8, focusing on chain ID indexing.
  
- **Bug Fixes**
	- Corrected handling of chain identifiers in various functions to ensure consistency and reliability.

- **Documentation**
	- Updated command descriptions and usage examples for querying chain nonces to reflect the new parameter naming conventions.
  
- **Chores**
	- Cleaned up deprecated references to `index` across the codebase to streamline functionality.

### CODE REVIEW

File: x/authority/migrations/v2/migrate.go [hunk 1]
```diff
@@ -15,7 +15,6 @@ func MigrateStore(
 	ctx sdk.Context,
 	keeper authorityKeeper,
 ) error {
-	ctx.Logger().Info("Migrating authority store from version 1 to 2")
```
Comment (lumtis) [inline x/authority/migrations/v2/migrate.go:18]: Logs already emitted by Cosmos SDK

File: proto/zetachain/zetacore/observer/chain_nonces.proto [hunk 2]
```diff
@@ -7,7 +7,10 @@ option go_package = "github.com/zeta-chain/zetacore/x/observer/types";
 
 message ChainNonces {
   string creator = 1;
+
+  // deprecated(v19): index has been replaced by chain_id for unique identifier
   string index = 2;
```
Comment (swift1337) [inline proto/zetachain/zetacore/observer/chain_nonces.proto:12]:
```suggestion
  string index = 2 [deprecated = true];;
```

File: x/observer/client/cli/query_chain_nonce.go [hunk 3]
```diff
@@ -45,16 +46,21 @@ func CmdListChainNonces() *cobra.Command {
...
 		Short: "shows a chainNonces",
 		Args:  cobra.ExactArgs(1),
 		RunE: func(cmd *cobra.Command, args []string) error {
 			clientCtx := client.GetClientContextFromCmd(cmd)
 
 			queryClient := types.NewQueryClient(clientCtx)
 
+			chainID, err := strconv.ParseInt(args[0], 10, 64)
+			if err != nil {
+				return err
```
Comment (swift1337) [inline x/observer/client/cli/query_chain_nonce.go:59]:
```suggestion
				return errors.Wrap(err, "unable to parse chain id from %q", args[0])
```

File: x/observer/migrations/v8/migrate.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+import (
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/zeta-chain/zetacore/x/observer/types"
+)
+
+type obsserverKeeper interface {
```
Comment (swift1337) [inline x/observer/migrations/v8/migrate.go:11]:
```suggestion
type observerKeeper interface {
```

File: x/observer/types/genesis.go [hunk 5]
```diff
@@ -38,13 +38,13 @@ func (gs GenesisState) Validate() error {
 	}
 
 	// Check for duplicated index in chainNonces
-	chainNoncesIndexMap := make(map[string]bool)
+	chainNoncesIndexMap := make(map[int64]bool)
```
Comment (swift1337) [inline x/observer/types/genesis.go:41]: nit: in Go you can use a map of `struct{}` to allocate less memory. In particular no memory is allocated for map values at all. Example: `map[int64]struct{}{}`

**Review by swift1337 — APPROVED** [review]
LGTM

File: x/observer/types/genesis.go (hunk 5 above)
Comment (lumtis) [reply x/observer/types/genesis.go:41]: [85eef7c](https://github.com/zeta-chain/node/pull/2515/commits/85eef7c5c3086048fdb1c1dc64cb0676f6ef1d1c)

File: x/observer/keeper/chain_nonces.go [hunk 6]
```diff
@@ -14,14 +16,14 @@ import (
 func (k Keeper) SetChainNonces(ctx sdk.Context, chainNonces types.ChainNonces) {
 	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.ChainNoncesKey))
 	b := k.cdc.MustMarshal(&chainNonces)
-	store.Set(types.KeyPrefix(chainNonces.Index), b)
+	store.Set(types.KeyPrefix(strconv.FormatInt(chainNonces.ChainId, 10)), b)
```
Comment (kingpinXD) [inline x/observer/keeper/chain_nonces.go:19]: Nit: we can add a function to create the Key from the chain id

File: x/observer/migrations/v8/migrate_test.go [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+		// set chain nonces
+		for _, chainNonce := range chainNoncesArray {
+			setChainNoncesLegacy(ctx, *k, chainNonce)
+		}
+
+		// there are 10 chain nonces in the store
+		allChainNonces := k.GetAllChainNonces(ctx)
+		require.Len(t, allChainNonces, 10)
+
+		// no chain nonces can be found in the store
```
Comment (kingpinXD) [inline x/observer/migrations/v8/migrate_test.go:81]:
```suggestion
		// no chain nonces can be found in the store using the new indexing
```

File: proto/zetachain/zetacore/observer/chain_nonces.proto [hunk 8]
```diff
@@ -7,7 +7,11 @@ option go_package = "github.com/zeta-chain/zetacore/x/observer/types";
 
 message ChainNonces {
   string creator = 1;
-  string index = 2;
+
+  // deprecated(v19): index has been replaced by chain_id for unique identifier
+  string index = 2 [ deprecated = true ];
+  ;
```
Comment (kingpinXD) [inline proto/zetachain/zetacore/observer/chain_nonces.proto:13]:
```suggestion
  
```

File: x/observer/keeper/chain_nonces.go (hunk 6 above)
Comment (lumtis) [reply x/observer/keeper/chain_nonces.go]: f8734e25e

**Review by kingpinXD — APPROVED** [review]

**Pull request merged**